corehq/apps/hqwebapp/management/commands/build_requirejs.py

```diff
@@ -205,3 +97,141 @@ def handle(self, **options):
+                                  "This script overwrites some static files. "
+                                  "Are you sure you want to continue (y/n)? ".format(out))
+        if confirm[0].lower() != 'y':
+            exit()
+    confirm = six.moves.input("You are running locally. Have you already run "
+                              "`./manage.py collectstatic --noinput && ./manage.py compilejsi18n` (y/n)? ")
+    if confirm[0].lower() != 'y':
+        exit()
+
+
+def _r_js(local=False, no_optimize=False):
+    '''
+    Write build.js file to feed to r.js, run r.js, and return filenames of the final build config
+    and the bundle config output by the build.
+    '''
+    with open(os.path.join(ROOT_DIR, 'staticfiles', 'hqwebapp', 'yaml', 'requirejs.yaml'), 'r') as f:
+        config = yaml.safe_load(f)
+
+    if no_optimize:
+        config['optimize'] = 'none'
+
+    html_files, local_js_dirs = _get_html_files_and_local_js_dirs(local)
+
+    # For each directory, add an optimized "module" entry including all of the main modules in that dir.
+    # For each of these entries, r.js will create an optimized bundle of these main modules and all their
+    # dependencies
+    dirs_to_js_modules = _get_main_js_modules_by_dir(html_files)
+    for directory, mains in dirs_to_js_modules.items():
+        config['modules'].append({
+            'name': os.path.join(directory, "bundle"),
+            'exclude': ['hqwebapp/js/common', 'hqwebapp/js/base_main'],
+            'include': sorted(mains),
+            'create': True,
+        })
+
+    _save_r_js_config(config)
+
+    call(["node", "bower_components/r.js/dist/r.js", "-o", BUILD_JS_FILENAME])
+
+    return config, local_js_dirs
+
+
+def _get_html_files_and_local_js_dirs(local):
+    """
+    Returns all HTML files in corehq, excluding partials
+    """
+    prefix = os.path.join(ROOT_DIR, 'corehq')
+    html_files = []
+    local_js_dirs = set()  # a reference of js filenames, for use when copying optimized bundles back into corehq
+    for root, dirs, files in os.walk(prefix):
+        for name in files:
+            if name.endswith(".html"):
+                filename = os.path.join(root, name)
+                if not re.search(r'/partials/', filename):
+                    html_files.append(filename)
+            elif local and name.endswith(".js"):
+                local_js_dirs.add(_relative(root))
+    return html_files, local_js_dirs
+
+
+def _get_main_js_modules_by_dir(html_files):
+    """
+    Returns a dict of all main js modules, grouped by directory:
+    {
+        'locations/js': set(['locations/js/import', 'locations/js/location', ...  ]),
+        'linked_domain/js': set(['linked_domain/js/domain_links']),
+        ...
+    }
+    """
+    dirs = defaultdict(set)
+    for filename in html_files:
+        proc = subprocess.Popen(["grep", r"^\s*{% requirejs_main [^%]* %}\s*$", filename],
```

Comment:
To be consistent with the first commit, should this use `with subprocess.Popen(...) as proc:`?